src/System.Private.CoreLib/shared/System/Text/Unicode/Utf8.cs

```diff
@@ -0,0 +1,182 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System.Buffers;
+using System.Diagnostics;
+
+namespace System.Text.Unicode
+{
+    public static class Utf8
+    {
+        /*
+         * OperationStatus-based APIs for transcoding of chunked data.
+         * This method is similar to Encoding.UTF8.GetBytes / GetChars but has a
+         * different calling convention, different error handling mechanisms, and
+         * different performance characteristics.
+         *
+         * If 'replaceInvalidSequences' is true, the method will replace any ill-formed
+         * subsequence in the source with U+FFFD when transcoding to the destination,
+         * then it will continue processing the remainder of the buffers. Otherwise
+         * the method will return OperationStatus.InvalidData.
+         *
+         * If the method does return an error code, the out parameters will represent
+         * how much of the data was successfully transcoded, and the location of the
+         * ill-formed subsequence can be deduced from these values.
+         *
+         * If 'replaceInvalidSequences' is true, the method is guaranteed never to return
+         * OperationStatus.InvalidData. If 'isFinalBlock' is true, the method is
+         * guaranteed never to return OperationStatus.NeedMoreData.
+         */
+
+        public static OperationStatus FromUtf16(ReadOnlySpan<char> source, Span<byte> destination, out int numCharsRead, out int numBytesWritten, bool replaceInvalidSequences = true, bool isFinalBlock = true)
+        {
+            int originalSourceLength = source.Length;
+            int originalDestinationLength = destination.Length;
+            OperationStatus status = OperationStatus.Done;
+
+            // In a loop, this is going to read and transcode one scalar value at a time
+            // from the source to the destination.
+
+            while (!source.IsEmpty)
+            {
+                status = Rune.DecodeUtf16(source, out Rune firstScalarValue, out int charsConsumed);
+
+                switch (status)
```

Comment:
> switch (status) [](start = 15, length = 16)

would it be faster if you check for success in the loop and then goto the switch statement only when not? or you think the jit is smart enough to optimize that?

Reply:
Yes, that would probably be faster. However, these implementations are going away in the near future once the fast-path comes online, so I'm not too concerned about temporary perf impact.

Reply:
(That is, the point of this PR is to get the basic API checked in and implemented correctly. The implementations aren't necessarily intended to be highly optimized just yet.)